reframe/core/hooks.py

```diff
@@ -0,0 +1,109 @@
+# Copyright 2016-2021 Swiss National Supercomputing Centre (CSCS/ETH Zurich)
+# ReFrame Project Developers. See the top-level LICENSE file for details.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+
+import functools
+
+import reframe.utility as util
+
+
+def _attach_hooks(hooks, name=None):
+    def _deco(func):
+        def select_hooks(obj, kind):
+            if name is None:
+                fn_name = func.__name__
+                if fn_name == '__init__':
+                    fn_name = 'init'
+
+                hook_kind = kind + fn_name
+            elif name is not None and name.startswith(kind):
+                hook_kind = name
+            else:
+                # Just any name that does not exist
+                hook_kind = 'xxx'
+
+            if hook_kind not in hooks:
+                return []
+
+            return [h for h in hooks[hook_kind]
+                    if h.name not in obj._disabled_hooks]
```

Comment:
```suggestion
def _attach_hooks(hooks, phase=None):
    def _deco(func):
        def select_hooks(obj, kind):
            if phase is None:
                fn_name = func.__name__
                if fn_name == '__init__':
                    fn_name = 'init'

                phase = kind + fn_name
            elif phase is not None and not phase.startswith(kind):
                # Just any name that does not exist
                phase = 'xxx'

            if phase not in hooks:
                return []

            return [h for h in hooks[phase]
                    if h.name not in obj._disabled_hooks]
```

Reply:
Renaming `name` and `hook_kind` as `phase` makes it consistent with the rest of the code and IMO makes this a lot easier to understand.

Reply:
I did it, except that I did not rename the `name` argument of `_attach_hooks`. You have to copy that in the `select_hooks`, because otherwise you would need to mark `phase` as `nonlocal`, which is not correct, because you will be sharing it across invocations of `select_hooks()` (and unit tests failed as a result).